File: Baekjun/8.3/1199. 오일러 회로/solution.py
import sys

n = int(input())

input = sys.stdin.readline
graph = []
degree = [0] * n
visit = [[0] * n for _ in range(n)]

for i in range(n):
    lst = {}

    for j, v in enumerate(list(map(int, input().split()))):
        if v:
            lst[j] = 1
            visit[i][j] = v
            degree[i] += v

    graph.append(lst)

for i in range(n):
    if degree[i] % 2:
        print(-1)
        sys.exit()

answer = []
stack = [0]

while stack:
    current = stack[-1]

    if graph[current]:
        _next = next(iter(graph[current]))
        visit[_next][current] -= 1
        visit[current][_next] -= 1
        degree[current] -= 1
        degree[_next] -= 1

        if not visit[current][_next]:
            del graph[current][_next]
            del graph[_next][current]

        stack.append(_next)
    else:
        answer.append(stack.pop() + 1)

for i in range(n):
    if degree[i]:
        print(-1)
        sys.exit()

print(" ".join(map(str, answer)))

File: Baekjun/8.3/1199. 오일러 회로/solution.swift
func solution() {
    let N = Int(readLine()!)!

    var degree = Array(repeating: 0, count: N)
    var visit = Array(repeating: Array(repeating: 0, count: N), count: N)
    var graph:[[Int:Int]] = []

    for i in 0..<N {
        var lst:[Int:Int] = [:]

        if let input = readLine() {
            let numbers = input.split(separator: " ").compactMap { Int($0) }

            for (idx, j) in numbers.enumerated() {
                if j != 0 {
                    degree[i] += j
                    visit[i][idx] = j
                    lst[idx] = 1
                }
            }
        }

        graph.append(lst)
    }

    for i in 0..<N {
        if degree[i] % 2 == 1 {
            print(-1)
            return
        }
    }

    var answer:[Int] = []
    var stack:[Int] = [0]

    while !stack.isEmpty {

        if let current = stack.last, !graph[current].isEmpty {
            let (key, _) = graph[current].first!

            visit[current][key] -= 1
            visit[key][current] -= 1
            degree[current] -= 1
            degree[key] -= 1

            if visit[current][key] == 0 {
                graph[current].removeValue(forKey: key)
                graph[key].removeValue(forKey: current)
            }

            stack.append(key)
        } else {
            answer.append(stack.popLast()! + 1)
        }
    }

    for i in 0..<N {
        if degree[i] > 0 {
            print(-1)
            return
        }
    }

    print(answer.map { String($0)}.joined(separator: " "))
}

solution()
